README
NAME
    App::Virtualenv - Perl virtual environment

VERSION
    version 2.07

SYNOPSIS
            #!/bin/sh
            perl -MApp::Virtualenv -erun -- environment_path

DESCRIPTION
    App::Virtualenv is a Perl package to create isolated Perl virtual
    environments, like Python virtual environment.

    See also: virtualenv.pl
    <https://metacpan.org/pod/distribution/App-Virtualenv/lib/App/Virtualenv
    /virtualenv.pl>

Functions
  sh(@args)
    runs shell program defined in SHELL environment variable, otherwise
    /bin/sh

    @args: *arguments of shell program*

    return value: *exit code of shell program*

  perl(@args)
    runs Perl interpreter

    @args: *arguments of Perl interpreter*

    return value: *exit code of Perl interpreter*

  activate($virtualenv_path)
    activates Perl virtual environment

    $virtualenv_path: *virtual environment path*

    return value: *virtual environment path if success, otherwise undef*

  deactivate($nondestructive)
    deactivates Perl virtual environment

    $nondestructive: *leaves envionment variables as it is, unless there are
    old envionment variables*

    return value: *always 1*

  create($virtualenv_path, $empty)
    creates Perl virtual environment

    $virtualenv_path: *new virtual environment path*

    $empty: *create empty virtual environment*

    return value: *virtual environment path if success, otherwise undef*

  find_virtualenv_path($virtualenv_path)
    finds Perl virtual environment path by $virtualenv_path argument or
    activated virtual environment or running script or PERL5LIB environment
    variable

    $virtualenv_path: *virtual environment path*

    return value: *best matching virtual environment path*

  activate2($virtualenv_path, $inform)
    activates Perl virtual environment by find_virtualenv_path function

    $virtualenv_path: *virtual environment path*

    $inform: *informs activated virtual environment path to STDERR if new
    activated path differs old one*

    return value: *activated best matching virtual environment path if
    success, otherwise undef*

  getinc($virtualenv_path)
    gets array ref of include paths given virtual environment path or
    sitelib paths

    $virtualenv_path: *virtual environment path*

    return value: *array ref of paths*

  list(%params)
    lists packages or modules or files by given %params

    %params: *parameters of function*

        one: *output is one-column, by default 0*

        detail: *prints additional detail by given value: module or file. by
        default undef*

    return value: *always 1*

  main(@argv)
    App::Virtualenv main function to run on command-line

    See also: virtualenv.pl
    <https://metacpan.org/pod/distribution/App-Virtualenv/lib/App/Virtualenv
    /virtualenv.pl>

    @argv: *command-line arguments*

    return value: *exit code of program*

  run
    runs App::Virtualenv by main function with command-line arguments by
    @ARGV

    return value: *function doesn't return, exits with main function return
    code*

PREVIOUS VERSION
    Previous version of App::Virtualenv has include PiV(Perl in Virtual
    environment) to list/install/uninstall modules using CPANPLUS API. Aimed
    with PiV making a package manager like Python pip. But Perl has various
    powerful package tools mainly CPAN and cpanminus, CPANPLUS and etc. And
    also building a great package manager requires huge community support.
    So, PiV is deprecated in version 2.xx.

    You should uninstall previous version before upgrading from v1.xx: cpanm
    -U App::Virtualenv; cpanm -i App::Virtualenv;

    See also: App::Virtualenv 1.13
    <https://metacpan.org/release/ORKUN/App-Virtualenv-1.13>

  Deprecated Modules
    *   App::Virtualenv::Piv

    *   App::Virtualenv::Module

    *   App::Virtualenv::Utils

INSTALLATION
    To install this module type the following

            perl Makefile.PL
            make
            make test
            make install

    from CPAN

            cpan -i App::Virtualenv

    You should uninstall previous version before upgrading from v1.xx: cpanm
    -U App::Virtualenv; cpanm -i App::Virtualenv;

DEPENDENCIES
    This module requires these other modules and libraries:

    *   local::lib

    *   ExtUtils::Installed

    *   CPAN

    *   Cwd

    *   Lazy::Utils

REPOSITORY
    GitHub <https://github.com/orkunkaraduman/perl5-virtualenv>

    CPAN <https://metacpan.org/release/App-Virtualenv>

SEE ALSO
    *   App::Virtualenv 1.13
        <https://metacpan.org/release/ORKUN/App-Virtualenv-1.13>

    *   CPAN <https://metacpan.org/pod/CPAN>

    *   App::cpanminus <https://metacpan.org/pod/App::cpanminus>

    *   CPANPLUS <https://metacpan.org/pod/CPANPLUS>

AUTHOR
    Orkun Karaduman (ORKUN) <orkun@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2017 Orkun Karaduman <orkunkaraduman@gmail.com>

    This program is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the License, or (at your
    option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program. If not, see <http://www.gnu.org/licenses/>.



Changes
2.07	2017-05-20
		- POD changes

2.06	2017-03-16
		- Accepting more than one packages when listing packages

2.05	2017-03-08
		- CPAN min required version setted 2.00
		- Improved creating virtual environment

2.04	2017-03-08
		- Script files bug fix for #! /usr/bin/perl
		- POD changes

2.03	2017-03-06
		- Perl v5.10.1 compatibility

2.02	2017-03-05
		- Fix for Experimental keys on scalar is now forbidden, Perl >= 5.24

2.01	2017-03-04
		- POD changes

2.00	2017-03-03
		- Removed Perl in Virtual environment (PiV)
		- App::Virtualenv::Piv is deprecated now
		- App::Virtualenv::Module is deprecated now
		- Bug fixed for finding best matching virtual environment
		- App::Virtualenv just use CPAN now
		- Added new features to virtualenv.pl
		- Improved documentation

1.13	2017-01-17
		- Added AUTOMATED_TESTING in Piv for noninteractive testing
		- Used PERL5LIB in Virtualenv::activate

1.12	2017-01-07
		- App::Virtualenv::Utils is deprecated
		- Noninteractive piv create/install/remove
		- Added piv -h
		- Documentation is improved

1.11	2016-12-15
		- App::Virtualenv::Module::inPerLib() fix
		- App::Virtualenv::activate2() added
		- Perl v5.14 for all modules

1.10	2016-12-10
		- Version, POD and README fixes for v1.09
		- Improved local library and Perl library module checks
		- Added piv install -v
		- Added piv remove -v
		- Added summary for piv install
		- Added summary for piv remove
		- Documentation is improved

1.09	2016-12-10
		- Improved local library and Perl library module checks
		- Added piv install -v
		- Added piv remove -v
		- Added summary for piv install
		- Added summary for piv remove
		- Documentation is improved

1.08	2016-12-04
		- Hotfix for versioning

1.07	2016-12-04
		- Minimum Perl version setted v5.14
		- Added piv install -t
		- Added piv install -s
		- Module.pm checking installed module is fixed
		- Module.pm $Config{sitearch} fix

1.06	2016-12-04
		- Added piv remove -f
		- Added piv virtualenv -e
		- Added archlib check for Perl library
		- Added archlib for installed

1.05	2016-12-03
		- Perl::Shell removed from project
		- Removed wrapper functions from App::Virtualenv
		- App::Virtualenv::Utils module added
		- App::Virtualenv::Module module ok with CPANPLUS
		- App::Virtualenv::Piv added

1.04	2016-11-29
		- Kwalitee Issue - has_abstract_in_pod fix
		- Added Change log
		- Improved modules
		- Improved Perl POD sections
		- Little bug fixes

1.03	2016-11-28
		- Module name changed to App::Virtualenv from App::virtualenv
		- Added Perl PODs
		- Added META_MERGE to Makefile.PL
		- Changed README.md
		- Initialized App::Virtualenv::Module but not implemented yet

1.02	2016-11-27
		- Added Perl::Shell in shell.pl
		- Bug fixes

1.01	2016-11-25
		- Version fixes
		- README & COPYRIGHT improvements

1.00	2016-11-23
		- First release


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;


WriteMakefile(
	NAME				=> 'App::Virtualenv',
	VERSION_FROM		=> 'lib/App/Virtualenv.pm',
	MIN_PERL_VERSION	=> '5.010001',
	PREREQ_PM			=> {
		'local::lib'	=> '2.000004',
		'ExtUtils::Installed' => '2.02',
		'CPAN'			=> '2.00',
		'Cwd'			=> '3.47',
		'Lazy::Utils'	=> '1.11',
	},
	EXE_FILES			=> [qw(
		lib/App/Virtualenv/virtualenv.pl
		lib/App/Virtualenv/sh.pl
		lib/App/Virtualenv/perl.pl
	)],
	AUTHOR				=> 'Orkun Karaduman <orkunkaraduman@gmail.com>',
	ABSTRACT			=> 'Perl virtual environment',
	LICENSE				=> 'gpl_3',
	META_MERGE			=> {
		'meta-spec'		=> {
			version		=> 2,
		},
		resources		=> {
			homepage	=> 'https://metacpan.org/release/App-Virtualenv',
			license		=> 'https://www.gnu.org/licenses/gpl-3.0.html',
			repository	=> {
				type	=> 'git',
				url		=> 'git://git@github.com:orkunkaraduman/perl5-virtualenv.git',
				web		=> 'https://github.com/orkunkaraduman/perl5-virtualenv',
			},
			bugtracker	=> {
				web		=> 'https://github.com/orkunkaraduman/perl5-virtualenv/issues',
				mailto	=> 'bug-App-Virtualenv@rt.cpan.org',
			},
		},
	},
);
